File: examples/leaflet-view/src/main/java/com/gurunars/leaflet_view/example/ActivityMain.java
package com.gurunars.leaflet_view.example;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.TextView;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.gurunars.leaflet_view.LeafletView;
import com.gurunars.leaflet_view.NoPageRenderer;
import com.gurunars.leaflet_view.PageRenderer;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;


public class ActivityMain extends AppCompatActivity {

    private List<TitledPage> pages = new ArrayList<>();
    private LeafletView<View, TitledPage> leafletView;

    private void updateAdapter() {
        Collections.sort(pages, new Comparator<TitledPage>() {
            @Override
            public int compare(TitledPage lhs, TitledPage rhs) {
                return lhs.getTitle().compareTo(rhs.getTitle());
            }
        });
        leafletView.setPages(pages);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        leafletView = (LeafletView<View, TitledPage>) findViewById(R.id.leafletView);
        leafletView.setPageRenderer(new PageRenderer<View, TitledPage>() {
            @Override
            public View renderPage(TitledPage page) {
                View viewGroup = LayoutInflater.from(ActivityMain.this).inflate(
                        R.layout.page_view, leafletView, false);
                TextView textView = (TextView) viewGroup.findViewById(
                        R.id.pageTitle);
                textView.setText(page.toString());
                viewGroup.setTag(page);
                return viewGroup;
            }

            @Override
            public void enter(View pageView) {
                setTitle(pageView.getTag().toString());
            }

            @Override
            public void leave(View pageView) {

            }
        });

        leafletView.setNoPageRenderer(new NoPageRenderer() {
            @Override
            public View renderNoPage() {
                return LayoutInflater.from(ActivityMain.this).
                        inflate(R.layout.no_page_view, leafletView, false);
            }

            @Override
            public void enter() {
                setTitle(R.string.empty);
            }

        });

        load();
        updateAdapter();
    }

    private void load() {
        pages = new Gson().fromJson(getPreferences(Context.MODE_PRIVATE).getString("data", "[]"), new TypeToken<ArrayList<TitledPage>>(){}.getType());
    }

    private void save() {
        SharedPreferences.Editor editor = getPreferences(Context.MODE_PRIVATE).edit();
        editor.putString("data", new Gson().toJson(pages));
        editor.apply();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.activity_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.create:
                createPage();
                return true;
            case R.id.delete:
                deletePage();
                return true;
            case R.id.edit:
                editPage();
                return true;
            case R.id.go_to:
                goToPage();
                return true;
            case R.id.clear:
                clear();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private static class TitledPageAdapter extends ArrayAdapter<TitledPage> {

        TitledPageAdapter(Context context, List<TitledPage> items) {
            super(context, 0, items);
        }

        @NonNull
        @Override
        public View getView(int position, View convertView, @NonNull ViewGroup parent) {
            if (convertView == null) {
                convertView = LayoutInflater.from(getContext()).inflate(android.R.layout.simple_list_item_1, parent, false);
                TextView text = (TextView) convertView.findViewById(android.R.id.text1);
                text.setText(getItem(position).getTitle());
            }
            return convertView;
        }

    }

    private void goToPage() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.go_to);
        builder.setSingleChoiceItems(
                new TitledPageAdapter(this, pages), -1,
                new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        leafletView.goTo(pages.get(which));
                        dialog.dismiss();
                    }
        });
        builder.setCancelable(true);
        builder.show();
    }

    private void editPage() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.edit);
        final EditText input = new EditText(this);
        input.setId(R.id.pageTitle);
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        TitledPage currentPage = leafletView.getCurrentPage();
        if (currentPage == null) {
            return;
        }
        input.setText(currentPage.getTitle());
        builder.setView(input);
        // Set up the buttons
        builder.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                TitledPage page = leafletView.getCurrentPage();
                page.setTitle(input.getText().toString());
                // NOTE: surely equals method could have been implemented
                // however the idea is to demo that these methods are not important -
                // only getId method is
                for (int i=0; i < pages.size(); i++) {
                    if (pages.get(i).getId() == page.getId()) {
                        pages.set(i, page);
                    }
                }
                updateAdapter();
                save();
            }
        });
        builder.show();
    }

    private void deletePage() {
        // NOTE: surely equals method could have been implemented
        // however the idea is to demo that these methods are not important - only getId method is
        TitledPage page = leafletView.getCurrentPage();
        for (int i=0; i < pages.size(); i++) {
            if (pages.get(i).getId() == page.getId()) {
                pages.remove(i);
            }
        }
        updateAdapter();
        save();
    }

    private void createPage() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.edit);
        final EditText input = new EditText(this);
        input.setId(R.id.pageTitle);
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        builder.setView(input);
        // Set up the buttons
        builder.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                TitledPage page = new TitledPage(input.getText().toString());
                pages.add(page);
                updateAdapter();
                save();
            }
        });
        builder.show();
    }

    private void clear() {
        pages.clear();
        updateAdapter();
        save();
    }

}

File: examples/leaflet-view/src/main/java/com/gurunars/leaflet_view/example/ActivityMain.kt
package com.gurunars.leaflet_view.example

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.InputType
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.TextView

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.gurunars.leaflet_view.LeafletView
import com.gurunars.leaflet_view.NoPageRenderer
import com.gurunars.leaflet_view.PageRenderer

import java.util.ArrayList
import java.util.Collections
import java.util.Comparator

import butterknife.ButterKnife


class ActivityMain : AppCompatActivity() {

    private var pages: MutableList<TitledPage> = ArrayList()
    private var leafletView: LeafletView<View, TitledPage>? = null

    private fun updateAdapter() {
        Collections.sort(pages) { lhs, rhs -> lhs.title.compareTo(rhs.title) }
        leafletView!!.setPages(pages)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        leafletView = ButterKnife.findById<LeafletView<View, TitledPage>>(this, R.id.leafletView)
        leafletView!!.setPageRenderer(object : PageRenderer<View, TitledPage> {
            fun renderPage(page: TitledPage): View {
                val viewGroup = LayoutInflater.from(this@ActivityMain).inflate(
                        R.layout.page_view, leafletView, false)
                val textView = viewGroup.findViewById(
                        R.id.pageTitle) as TextView
                textView.text = page.toString()
                viewGroup.tag = page
                return viewGroup
            }

            override fun enter(pageView: View) {
                title = pageView.tag.toString()
            }

            override fun leave(pageView: View) {

            }
        })

        leafletView!!.setNoPageRenderer(object : NoPageRenderer {
            override fun renderNoPage(): View {
                return LayoutInflater.from(this@ActivityMain).inflate(R.layout.no_page_view, leafletView, false)
            }

            override fun enter() {
                setTitle(R.string.empty)
            }

        })

        load()
        updateAdapter()
    }

    private fun load() {
        pages = Gson().fromJson<List<TitledPage>>(getPreferences(Context.MODE_PRIVATE).getString("data", "[]"), object : TypeToken<ArrayList<TitledPage>>() {

        }.type)
    }

    private fun save() {
        val editor = getPreferences(Context.MODE_PRIVATE).edit()
        editor.putString("data", Gson().toJson(pages))
        editor.apply()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.activity_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        when (item.itemId) {
            R.id.create -> {
                createPage()
                return true
            }
            R.id.delete -> {
                deletePage()
                return true
            }
            R.id.edit -> {
                editPage()
                return true
            }
            R.id.go_to -> {
                goToPage()
                return true
            }
            R.id.clear -> {
                clear()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    private class TitledPageAdapter internal constructor(context: Context, items: List<TitledPage>) : ArrayAdapter<TitledPage>(context, 0, items) {

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var convertView = convertView
            if (convertView == null) {
                convertView = LayoutInflater.from(context).inflate(android.R.layout.simple_list_item_1, parent, false)
                val text = convertView!!.findViewById(android.R.id.text1) as TextView
                text.text = getItem(position)!!.title
            }
            return convertView
        }

    }

    private fun goToPage() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.go_to)
        builder.setSingleChoiceItems(
                TitledPageAdapter(this, pages), -1
        ) { dialog, which ->
            leafletView!!.goTo(pages[which])
            dialog.dismiss()
        }
        builder.setCancelable(true)
        builder.show()
    }

    private fun editPage() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.edit)
        val input = EditText(this)
        input.id = R.id.pageTitle
        input.inputType = InputType.TYPE_CLASS_TEXT
        val currentPage = leafletView!!.getCurrentPage() ?: return
        input.setText(currentPage.title)
        builder.setView(input)
        // Set up the buttons
        builder.setPositiveButton(R.string.ok) { dialog, which ->
            val page = leafletView!!.getCurrentPage()
            page!!.title = input.text.toString()
            // NOTE: surely equals method could have been implemented
            // however the idea is to demo that these methods are not important -
            // only getId method is
            for (i in pages.indices) {
                if (pages[i].id == page!!.id) {
                    pages[i] = page
                }
            }
            updateAdapter()
            save()
        }
        builder.show()
    }

    private fun deletePage() {
        // NOTE: surely equals method could have been implemented
        // however the idea is to demo that these methods are not important - only getId method is
        val page = leafletView!!.getCurrentPage()
        for (i in pages.indices) {
            if (pages[i].id == page!!.id) {
                pages.removeAt(i)
            }
        }
        updateAdapter()
        save()
    }

    private fun createPage() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.edit)
        val input = EditText(this)
        input.id = R.id.pageTitle
        input.inputType = InputType.TYPE_CLASS_TEXT
        builder.setView(input)
        // Set up the buttons
        builder.setPositiveButton(R.string.ok) { dialog, which ->
            val page = TitledPage(input.text.toString())
            pages.add(page)
            updateAdapter()
            save()
        }
        builder.show()
    }

    private fun clear() {
        pages.clear()
        updateAdapter()
        save()
    }

}